presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,285 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multiset;
+import io.airlift.log.Logger;
+import io.airlift.stats.CounterStat;
+import io.airlift.units.Duration;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.raptor.storage.BucketBalancer.BucketAssignment.assignment;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * Service to balance buckets across active Raptor storage nodes in a cluster.
+ *
+ * - For a given distribution, each node should be allocated the same number of buckets. This enhances parallelism, and therefore query performance.
+ * - The total disk utilization across the cluster should be balanced. This ensures that total cluster storage capacity is maximized. Simply allocating the same number of buckets
+ *   to every node may not achieve this, as bucket sizes may vary dramatically across distributions.
+ *
+ * This prioritizes query performance over total cluster storage capacity.
+ */
+public class BucketBalancer
+{
+    private static final Logger log = Logger.get(BucketBalancer.class);
+
+    private final NodeSupplier nodeSupplier;
+    private final ShardManager shardManager;
+    private final Duration interval;
+    private final boolean backupAvailable;
+    private final boolean coordinator;
+    private final ScheduledExecutorService executor;
+
+    private final AtomicBoolean started = new AtomicBoolean();
+
+    private final CounterStat bucketsBalanced = new CounterStat();
+    private final CounterStat jobErrors = new CounterStat();
+
+    @Inject
+    public BucketBalancer(
+            NodeManager nodeManager,
+            NodeSupplier nodeSupplier,
+            ShardManager shardManager,
+            StorageManagerConfig config,
+            BackupService backupService,
+            RaptorConnectorId connectorId)
+    {
+        this(nodeSupplier,
+                shardManager,
+                config.getBalancerInterval(),
+                backupService.isBackupAvailable(),
+                nodeManager.getCurrentNode().isCoordinator(),
+                connectorId.toString());
+    }
+
+    public BucketBalancer(
```

Comment:
Can you decouple the balancer *service* from the balancing *logic*? Balancing is a pure function `[Node] -> Assignments -> Assignments` (we need to pass the nodes to have nodes that are not in the original assignments) with some properties (targetMin <= count <= targetMax for each node) and invariants (sum of bucket count).

This would help to build a CLI tool that uses the logic and also monitor the deviation of the bucket distribution against the better (we hope *optimal*) one.